File: atcoder/ABC/1/104ABC-C.py
d, g = map(int, input().split())
p = []
c = []
for _ in range(d):
    px, cx = map(int, input().split())
    p.append(px)
    c.append(cx)

ans = 2 << 9

# '1<<d'は1を2進数で2桁左にずらす
# 2進数を使ってるのは、MAXで2^D通り試せばいいからか
for m in range(1 << d):
    s = 0
    num = 0
    rest_max = -1
    for i in range(d):
        # mをi桁右にずらす（2進数）
        # &は論理積. 両方が1の時だけ1になる
        # コンプリートボーナスをとるかとらないかを全パターン網羅でやってるだけやな
        if m >> i & 1:
            s += 100 * (i + 1) * p[i] + c[i]
            num += p[i]
        else:
            # このrest_maxは、問題番号だから、まだコンプリートしていない最大の問題番号か
            # iは昇順でループするので、ここでrest_maxを更新するのは問題なし
            rest_max = i
    # まだ未達
    if s < g:
        s1 = 100 * (rest_max + 1)
        # 残っている最大の問題番号で達成できるか？
        need = (g - s + s1 - 1) // s1
        # 達成に必要な問題数が最大問題番号の問題数以上だったらcontinue
        # （コンプリート獲得なしでも達成できる数じゃなかったら）
        # これはコンプリートしない問題番号は1つのみで、2つ以上は存在しないことにしてるが、それでよいのか？反証はないか？
        # いや、よいな。複数の問題をコンプリートせずに半端にとくなら、最大の問題番号をコンプリートする場合の方が効率良いもん
        if need >= p[rest_max]:
            continue
        num += need
    # これまでの結果で最小の値を探す
    ans = min(ans, num)
print(ans)

File: atcoder/ABC/1/104ABC-C.cpp
#include <iostream>
#include <cmath>
using namespace std;
using ll = long long;

int main()
{
  ll D, G;

  cin >> D >> G;

  int P[D], C[D];
  for (int i = 0; i < D; i++)
  {
    cin >> P[i] >> C[i];
  }

  int ans = pow(10, 9);

  for (int i = 0; i < pow(2, D); i++)
  {
    int score = 0;
    int cnt = 0;
    int max_q_no = 0;
    // 何番目の問をコンプリートするか
    for (int j = 0; j < D; j++)
    {
      // 問題の数だけ2進数を使った判定を行う。2の階乗での全探索になってるので、指数が増えるごとに2進数の桁数も合わせて増える
      // そこに各桁で1が立ってるかどうかを1<<(問題番号-1)を使った論理積で判定する。
      // 2の問題数乗になっているので、問題番号における01と2進数の各桁の01を対応させられる。
      // ここでどの問をコンプリートするかどうかの判定
      // 1<<jを使うとj桁分左にずらした２進数の値が得られるので、1(1), 10(2), 100(4), 1000(8), 10000(16)
      // &で論理積をとることで、問題番号と桁番号を紐付けて01の判定を行える.
      // &を使うとiも2進数変換された上で論理積がとられる
      if (i & (1 << j))
      {
        score += 100 * (j + 1) * P[j] + C[j];
        cnt += P[j];
      }
      // コンプリートしない場合
      else
      {
        max_q_no = j;
      }
    }
    // まだ点数が足りない場合
    if (score < G)
    {
      // 残り
      int remain = G - score;
      // 最大の問題番号をコンプリートしない数だけ解いて達成できるか？
      int q_p = 100 * (max_q_no + 1);
      int need = ceil((remain + (q_p - 1)) / q_p);
      if (need < P[max_q_no])
      {
        cnt += need;
      }
      else
      {
        continue;
      }
    }

    ans = min(ans, cnt);
  }

  cout << ans;

  return 0;
}